src/udev/udevd.c

```diff
@@ -1159,56 +1159,54 @@ static int on_ctrl_msg(struct udev_ctrl *uctrl, enum udev_ctrl_msg_type type, co
         return 1;
 }
 
-static int synthesize_change_one(sd_device *dev, const char *syspath) {
-        const char *filename;
+static int synthesize_change_one(sd_device *dev, sd_device *target) {
         int r;
 
-        filename = strjoina(syspath, "/uevent");
-        log_device_debug(dev, "device is closed, synthesising 'change' on %s", syspath);
-        r = write_string_file(filename, "change", WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (DEBUG_LOGGING) {
+                const char *syspath = NULL;
+                (void) sd_device_get_syspath(target, &syspath);
+                log_device_debug(dev, "device is closed, synthesising 'change' on %s", strna(syspath));
+        }
+
+        r = sd_device_set_sysattr_value(target, "uevent", "change");
```

Comment:
This feels like a violation of semantics. `sd_device_set_sysattr_value` is for setting of attributes, and caches the attribute. That fact that `uevent` is a bit like an attribute is an accident of implementation. If anything, I think it'd be clearer to add a small helper to write to `uvent`, and not reuse `sd_device_set_sysattr_value` for this.

@poettering ?

Reply:
Some programs already use `sd_device_set_sysattr_value()` for writing `uevent`. So, if we refuse to use that, then we need to introduce `sd_device_set_action()` or so.

Reply:
Hmm, https://github.com/opensvc/multipath-tools/blob/66791b63c3ce5feb04cc617c039f060c4edf0988/multipathd/main.c#L1523-L1525

Reply:
Google says `sd_device_set_sysattr_value()` is not used with `"uevent"`.

Reply:
maybe we should actually have new call sd_device_fire() or so, that fires an uevent off, and encapsulates this? I think that would be pretty nice to have, the file is after all indeed not a regular file but something special.

(or sd_device_emit()? sd_device_uevent()? sd_device_trigger()? sd_device_trigger_uevent()? ideas? – i like shorter better, maybe just sd_device_trigger() given that "udevadm trigger" is ultimately just a big wrapper around this in the end)